modules/core/src/main/java/org/apache/ignite/internal/cdc/IgniteCDC.java

```diff
@@ -0,0 +1,485 @@
+
+            return log;
+        }
+        catch (IgniteCheckedException e) {
+            throw new IgniteException(e);
+        }
+    }
+
+    /**
+     * Resolves work directory.
+     *
+     * @return Working directory
+     */
+    public static String workDir(IgniteConfiguration icfg) {
+        try {
+            String igniteHome = icfg.getIgniteHome();
+
+            // Set Ignite home.
+            if (igniteHome == null)
+                igniteHome = U.getIgniteHome();
+            else
+                // If user provided IGNITE_HOME - set it as a system property.
+                U.setIgniteHome(igniteHome);
+
+            String userProvidedWorkDir = icfg.getWorkDirectory();
+
+            // Correctly resolve work directory and set it back to configuration.
+            return U.workDirectory(userProvidedWorkDir, igniteHome);
+        }
+        catch (IgniteCheckedException e) {
+            throw new IgniteException(e);
+        }
+    }
+
+    /**
+     * @param segment WAL segment file.
+     * @return Segment index.
+     */
+    public long segmentIndex(Path segment) {
+        String fn = segment.getFileName().toString();
+
+        return Long.parseLong(fn.substring(0, fn.indexOf('.')));
+    }
+
+    /**
+     * @param icfg Configuration.
+     * @return Consistent id to use.
+     */
+    private String consistentId(IgniteConfiguration icfg) {
+        if (icfg.getConsistentId() != null)
+            return U.maskForFileName(icfg.getConsistentId().toString());
+
+        if (cdcCfg.getAutoGeneratedConsistentId() == null) {
+            log.warning("ConsitentId is null.");
+
+            return null;
+        }
+
+        return PdsConsistentIdProcessor.genNewStyleSubfolderName(cdcCfg.getNodeIndex(), cdcCfg.getAutoGeneratedConsistentId());
+    }
+
+    /**
+     * Waits for the files or directories to be created insied {@code watchDir}
+     * and if new file pass the {@code filter} then {@code callback} notified with the newly create file.
+     * {@code callback} will allso be notified about already existing files that passes the filter.
+     *
+     * @param watchDir Directory to watch.
+     * @param filter Filter of events.
+     * @param sorter Sorter of files.
+     * @param callback Callback to be notified.
+     */
+    @SuppressWarnings("BusyWait")
+    public static void waitFor(Path watchDir, Predicate<Path> filter, Comparator<Path> sorter,
+        Predicate<Path> callback, long timeout, IgniteLogger log) throws InterruptedException {
+        // If watch dir not exists waiting for it creation.
+        if (!Files.exists(watchDir))
+            waitFor(watchDir.getParent(), watchDir::equals, Path::compareTo, p -> false, timeout, log);
+
+        try {
+            // Clear deleted file.
+            Set<Path> seen = new HashSet<>();
+
+            while (true) {
+                try (Stream<Path> children = Files.walk(watchDir, 1).filter(p -> !p.equals(watchDir))) {
```

Comment:
It may fail in case some of the files will be concurrently deleted from the directory. It is not a critical issue when only single application use this directory, however, I think it can be fixed in this PR.

Reply:
> It may fail in case some of the files will be concurrently deleted from the directory

It MUST fail if some application concurrently deletes segments from the directory.
Only one application can scan and process new segments from the CDC directory.
That's why we need a lock file at the start.